week2/assignment/main.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"sync"
+)
+
+func readFileByLine(filePath string, lines chan<-string, processingFile <-chan string, wg *sync.WaitGroup) {	
+	file, err := os.Open(filePath)
+	
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer func() {
+		file.Close();
+		
+		<- processingFile
+		if len(processingFile) == 0 {
+			close(lines)
+		}
+
+		wg.Done()
+    }()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		lines <- scanner.Text()		
+	}
+}
+
+func countWords(lines chan string, wordCount map[string]int, wg *sync.WaitGroup)  {
+	for line := range lines {
+		words := strings.Fields(line)
+
+		for _, word := range words { 
+			wordCount[word]++
+		}
+	}
+
+	defer wg.Done()
```

Comment:
In my opinion, `defer` should be at the beginning of this func. Because `defer` is always called when a function exits, putting it at the beginning would prevent us from forgetting.